cmd/find.go

```diff
@@ -0,0 +1,468 @@
+	errorIf(err, "Unable to construct client")
+	fileContent := parseContent(content)
+
+	if noAlias {
+		//pass hostname (alias) to to watchEvents so that it can remove from pathnames returned from watch
+		watchEvents(ctx, clnt, params, alias, fileContent)
+		return
+	}
+	watchEvents(ctx, clnt, params, "", fileContent)
+	return
+}
+
+// doFindOlder checks to see if the given object was created before or after the given time at which an object was created
+func doFindOlder(createTime time.Time, pattern string) bool {
+	i, err := TimeHelper(pattern)
+	now := time.Now()
+	fatalIf(probe.NewError(err), "Error parsing string passed to flag older")
+
+	//find all time in which the time in which the object was just created is after the current time
+	t := time.Date(now.Year(), now.Month(), now.Day()-i, now.Hour(), now.Minute(), 0, 0, time.UTC)
+	return createTime.Before(t)
+}
+
+// doFindNewer checks to see if the given object was created before the given threshold
+func doFindNewer(createTime time.Time, pattern string) bool {
+	i, err := TimeHelper(pattern)
+	now := time.Now()
+
+	fatalIf(probe.NewError(err), "Error parsing string passed to flag newer")
+
+	t := time.Date(now.Year(), now.Month(), now.Day()-i, now.Hour(), now.Minute(), 0, 0, time.UTC)
+	return createTime.After(t) || createTime.Equal(t)
+}
+
+// doFindLargerSize checks to see if the given object is larger than the given threshold
+func doFindLargerSize(size int64, pattern string) bool {
+	i, err := humanize.ParseBytes(pattern)
+	fatalIf(probe.NewError(err), "Error parsing string passed to flag larger")
+
+	return int64(i) < size
+}
+
+// doFindSmallerSize checks to see if the given object is smaller than the given threshold
+func doFindSmallerSize(size int64, pattern string) bool {
+	i, err := humanize.ParseBytes(pattern)
+	fatalIf(probe.NewError(err), "Error parsing string passed to flag smaller")
+
+	return int64(i) > size
+}
+
+// DoFind is used to handle most of the users input
+func DoFind(clnt Client, ctx *cli.Context) {
+	pathnameParts := strings.SplitAfter(ctx.Args().Get(0), "/")
+	alias := strings.TrimSuffix(pathnameParts[0], "/")
+	_, err := getHostConfig(alias)
+
+	// iterate over all content which is within the given directory
+	for content := range clnt.List(true, false, DirNone) {
+		fileContent := parseContent(content)
+		filePath := fileContent.Key
+
+		// traversing in a object store not a file path
+		if err == nil {
+			filePath = path.Join(alias, filePath)
+		}
+
+		if ctx.String("maxdepth") != "" {
+			i, e := strconv.Atoi(ctx.String("maxdepth"))
+			s := ""
+
+			fatalIf(probe.NewError(e), "Error parsing string passed to flag maxdepth")
+
+			// we are going to be parsing the path by x amounts
+			pathParts := strings.SplitAfter(filePath, "/")
+
+			if len(pathParts) < i || i < 0 {
+				i = len(pathParts)
+			}
+
+			// append portions of path into a string
+			for j := 0; j < i; j++ {
+				s += pathParts[j]
+			}
+
+			filePath = s
+			fileContent.Key = s
+		}
```

Comment:
There appears to be another corner case bug:

With find (executed in the mc codebase top directory):

```
$ find vendor/ -maxdepth 1
vendor/
vendor/golang.org
vendor/vendor.json
vendor/github.com
vendor/gopkg.in
$ find vendor/ -maxdepth 2
vendor/
vendor/golang.org
vendor/golang.org/x
vendor/vendor.json
vendor/github.com
vendor/github.com/dustin
vendor/github.com/rjeczalik
vendor/github.com/fatih
vendor/github.com/hashicorp
vendor/github.com/minio
vendor/github.com/cheggaaa
vendor/github.com/go-ini
vendor/github.com/mattn
vendor/github.com/pkg
vendor/gopkg.in
vendor/gopkg.in/check.v1
```

whereas with mc find:

```
$ mc find vendor/ --maxdepth 1
$ mc find vendor/ --maxdepth 2
vendor/vendor.json

```

`mc` is not listing directories - that is fine. However, with a max-depth of 1 it is not traversing the given path at all. There is an off-by-one bug here.